File: app/store.py
from __future__ import annotations

import json
import logging
import os
import threading
from dataclasses import dataclass, field
from datetime import datetime
from pathlib import Path
from typing import Any

from app import clock as clock_mod
from app import idgen
from app.config import Settings
from app.constants import TYPE_CLIENT_DETAILS
from app.kafka_publisher import KafkaPublisher, NullPublisher, Publisher
from app.models import ClientView, Event, RegisterRequest, RegisterResponse
from app.redis_live import RedisLive

LOG = logging.getLogger("trustedge-agent-api")


@dataclass
class DeviceRecord:
    device_id: str
    device_token: str = ""
    last_details: dict[str, Any] = field(default_factory=dict)
    last_seen_at: datetime | None = None

    def to_dict(self) -> dict[str, Any]:
        out: dict[str, Any] = {
            "device_id": self.device_id,
            "device_token": self.device_token,
        }
        if self.last_details:
            out["last_details"] = self.last_details
        if self.last_seen_at is not None:
            out["last_seen_at"] = self.last_seen_at.isoformat().replace("+00:00", "Z")
        return out


class EventStore:
    def __init__(
        self,
        settings: Settings,
        *,
        redis_live: RedisLive | None = None,
        publisher: Publisher | None = None,
    ) -> None:
        self._settings = settings
        self._max_events = settings.max_events if settings.max_events > 0 else 500
        self._disable_disk = not settings.persist_files()
        self._data_dir = Path(settings.data_dir)
        self._lock = threading.RLock()
        self._devices: dict[str, DeviceRecord] = {}
        self._tokens: dict[str, str] = {}
        self._events: dict[str, list[Event]] = {}
        self._redis = redis_live
        self._publisher = publisher or NullPublisher()

        if not self._disable_disk:
            self._data_dir.mkdir(parents=True, exist_ok=True)
            self._load_disk()
        if self._redis is not None and self._disable_disk:
            self._load_from_redis()

    @classmethod
    def from_settings(cls, settings: Settings) -> EventStore:
        redis_live: RedisLive | None = None
        if settings.redis_url.strip():
            redis_live = RedisLive(settings.redis_url.strip(), settings.max_events)
        publisher: Publisher = NullPublisher()
        if settings.kafka_brokers.strip():
            publisher = KafkaPublisher(settings.kafka_brokers.strip(), settings.kafka_topic)
        return cls(settings, redis_live=redis_live, publisher=publisher)

    @property
    def redis_enabled(self) -> bool:
        return self._redis is not None

    @property
    def kafka_enabled(self) -> bool:
        return not isinstance(self._publisher, NullPublisher)

    def close(self) -> None:
        if self._redis is not None:
            self._redis.close()
        self._publisher.close()

    def _load_disk(self) -> None:
        devices_path = self._data_dir / "devices.json"
        if devices_path.exists():
            data = json.loads(devices_path.read_text(encoding="utf-8"))
            if isinstance(data, list):
                for item in data:
                    if not isinstance(item, dict):
                        continue
                    rec = self._record_from_dict(item)
                    if rec.device_id:
                        self._devices[rec.device_id] = rec
                        if rec.device_token:
                            self._tokens[rec.device_token] = rec.device_id

        events_path = self._data_dir / "events.jsonl"
        if events_path.exists():
            for line in events_path.read_text(encoding="utf-8").splitlines():
                if not line.strip():
                    continue
                try:
                    payload = json.loads(line)
                    event = Event.model_validate(payload)
                except (json.JSONDecodeError, ValueError):
                    continue
                if not event.device_id:
                    continue
                events = self._events.setdefault(event.device_id, [])
                events.append(event)
                if len(events) > self._max_events:
                    self._events[event.device_id] = events[-self._max_events :]

    def _load_from_redis(self) -> None:
        if self._redis is None:
            return
        for item in self._redis.load_device_auth():
            rec = self._record_from_dict(item)
            if not rec.device_id:
                continue
            self._devices[rec.device_id] = rec
            if rec.device_token:
                self._tokens[rec.device_token] = rec.device_id

    @staticmethod
    def _record_from_dict(data: dict[str, Any]) -> DeviceRecord:
        last_seen = data.get("last_seen_at")
        parsed_seen: datetime | None = None
        if isinstance(last_seen, str) and last_seen:
            text = last_seen[:-1] + "+00:00" if last_seen.endswith("Z") else last_seen
            try:
                parsed_seen = datetime.fromisoformat(text)
            except ValueError:
                parsed_seen = None
        details = data.get("last_details")
        return DeviceRecord(
            device_id=str(data.get("device_id") or ""),
            device_token=str(data.get("device_token") or ""),
            last_details=details if isinstance(details, dict) else {},
            last_seen_at=parsed_seen,
        )

    def _persist_devices(self) -> None:
        if self._disable_disk:
            return
        records = [rec.to_dict() for rec in self._devices.values()]
        path = self._data_dir / "devices.json"
        path.write_text(json.dumps(records, indent=2) + "\n", encoding="utf-8")
        os.chmod(path, 0o600)

    def _append_event(self, event: Event) -> None:
        if self._disable_disk:
            return
        path = self._data_dir / "events.jsonl"
        with path.open("a", encoding="utf-8") as handle:
            handle.write(json.dumps(event.model_dump(mode="json", by_alias=False)) + "\n")
        os.chmod(path, 0o600)

    def register(self, req: RegisterRequest) -> RegisterResponse:
        with self._lock:
            device_id = req.device_id or idgen.new_device_id()
            rec = self._devices.get(device_id)
            if rec is None:
                rec = DeviceRecord(device_id=device_id)
                self._devices[device_id] = rec
            if not rec.device_token:
                rec.device_token = idgen.new_token()
                self._tokens[rec.device_token] = device_id

            now = clock_mod.now_utc()
            rec.last_seen_at = now
            if req.hostname:
                rec.last_details["hostname"] = req.hostname
            if req.os:
                rec.last_details["os"] = req.os
            if req.os_version:
                rec.last_details["os_version"] = req.os_version
            if req.arch:
                rec.last_details["arch"] = req.arch
            if req.agent_version:
                rec.last_details["agent_version"] = req.agent_version

            self._persist_devices()
            if self._redis is not None:
                self._redis.upsert_register(device_id, rec.last_details, now)
                if self._disable_disk:
                    self._redis.save_device_auth(rec.to_dict())

            return RegisterResponse(device_id=device_id, device_token=rec.device_token)

    def device_id_for_token(self, token: str) -> str | None:
        with self._lock:
            return self._tokens.get(token)

    def add_event(self, event: Event) -> None:
        with self._lock:
            rec = self._devices.get(event.device_id)
            if rec is None:
                rec = DeviceRecord(device_id=event.device_id)
                self._devices[event.device_id] = rec

            ts = event.ts or clock_mod.now_utc()
            event.ts = ts
            rec.last_seen_at = ts
            if event.type == TYPE_CLIENT_DETAILS:
                rec.last_details = dict(event.payload)

            events = self._events.setdefault(event.device_id, [])
            events.append(event)
            if len(events) > self._max_events:
                events = events[-self._max_events :]
            self._events[event.device_id] = events

            self._append_event(event)
            self._persist_devices()
            if self._redis is not None:
                self._redis.upsert_event(event)
                if self._disable_disk:
                    self._redis.save_device_auth(rec.to_dict())
            self._publisher.publish_event(event)

    def get_client(self, device_id: str, limit: int = 50) -> ClientView | None:
        with self._lock:
            rec = self._devices.get(device_id)
            if rec is None:
                return None
            if limit <= 0:
                limit = 50
            events = list(self._events.get(device_id, []))
            if len(events) > limit:
                events = events[-limit:]
            return ClientView(
                device_id=rec.device_id,
                last_details=rec.last_details or None,
                last_seen_at=rec.last_seen_at,
                recent_events=events,
            )

File: internal/store/store.go
package store

import (
	"encoding/json"
	"log"
	"os"
	"path/filepath"
	"sync"
	"time"

	"github.com/TrustEdgeOrg/TrustEdge-Agent-API/internal/clock"
	"github.com/TrustEdgeOrg/TrustEdge-Agent-API/internal/constants"
	"github.com/TrustEdgeOrg/TrustEdge-Agent-API/internal/kafka"
	"github.com/TrustEdgeOrg/TrustEdge-Agent-API/internal/models"
	"github.com/TrustEdgeOrg/TrustEdge-Agent-API/internal/idgen"
)

type deviceRecord struct {
	DeviceID    string         `json:"device_id"`
	DeviceToken string         `json:"device_token"`
	LastDetails map[string]any `json:"last_details,omitempty"`
	LastSeenAt  *time.Time     `json:"last_seen_at,omitempty"`
}

type Store struct {
	mu           sync.RWMutex
	clock        clock.Clock
	dataDir      string
	maxEvents    int
	disableDisk  bool
	devices      map[string]*deviceRecord
	tokens       map[string]string
	events       map[string][]models.Event
	live         *redisLive
	publisher    kafka.Publisher
}

type Options struct {
	Clock        clock.Clock
	DataDir      string
	MaxEvents    int
	// DisableDiskPersistence skips devices.json and events.jsonl (default: write to disk).
	DisableDiskPersistence bool
	RedisURL     string
	KafkaBrokers string
	KafkaTopic   string
	Logger       *log.Logger
	Publisher    kafka.Publisher
}

func New(dataDir string, maxEvents int) (*Store, error) {
	return NewWithOptions(Options{DataDir: dataDir, MaxEvents: maxEvents})
}

func NewWithOptions(opts Options) (*Store, error) {
	if opts.MaxEvents <= 0 {
		opts.MaxEvents = 500
	}
	if opts.Clock == nil {
		opts.Clock = clock.Real{}
	}
	s := &Store{
		clock:        opts.Clock,
		dataDir:      opts.DataDir,
		maxEvents:    opts.MaxEvents,
		disableDisk:  opts.DisableDiskPersistence,
		devices:      map[string]*deviceRecord{},
		tokens:       map[string]string{},
		events:       map[string][]models.Event{},
	}
	if !s.disableDisk {
		if err := os.MkdirAll(opts.DataDir, 0o755); err != nil {
			return nil, err
		}
		if err := s.load(); err != nil {
			return nil, err
		}
	}
	if opts.RedisURL != "" {
		live, err := newRedisLive(opts.RedisURL, opts.MaxEvents, opts.Logger, opts.Clock)
		if err != nil {
			return nil, err
		}
		s.live = live
		if s.disableDisk {
			if err := s.loadFromRedis(); err != nil {
				return nil, err
			}
		}
	}
	if opts.Publisher != nil {
		s.publisher = opts.Publisher
	} else if opts.KafkaBrokers != "" {
		pub, err := kafka.NewProducer(opts.KafkaBrokers, opts.KafkaTopic, opts.Logger)
		if err != nil {
			return nil, err
		}
		s.publisher = pub
	}
	return s, nil
}

func (s *Store) RedisEnabled() bool {
	return s.live != nil
}

func (s *Store) KafkaEnabled() bool {
	return s.publisher != nil
}

func (s *Store) Close() error {
	var err error
	if s.live != nil {
		err = s.live.Close()
	}
	if s.publisher != nil {
		if closeErr := s.publisher.Close(); closeErr != nil && err == nil {
			err = closeErr
		}
	}
	return err
}

func (s *Store) load() error {
	path := filepath.Join(s.dataDir, "devices.json")
	data, err := os.ReadFile(path)
	if err != nil {
		if os.IsNotExist(err) {
			return nil
		}
		return err
	}
	var devices []*deviceRecord
	if err := json.Unmarshal(data, &devices); err != nil {
		return err
	}
	for _, d := range devices {
		s.devices[d.DeviceID] = d
		if d.DeviceToken != "" {
			s.tokens[d.DeviceToken] = d.DeviceID
		}
	}
	eventsPath := filepath.Join(s.dataDir, "events.jsonl")
	raw, err := os.ReadFile(eventsPath)
	if err != nil {
		if os.IsNotExist(err) {
			return nil
		}
		return err
	}
	for _, line := range splitLines(raw) {
		if len(line) == 0 {
			continue
		}
		var ev models.Event
		if err := json.Unmarshal(line, &ev); err != nil {
			continue
		}
		s.events[ev.DeviceID] = append(s.events[ev.DeviceID], ev)
	}
	for id, list := range s.events {
		if len(list) > s.maxEvents {
			s.events[id] = list[len(list)-s.maxEvents:]
		}
	}
	return nil
}

func (s *Store) loadFromRedis() error {
	if s.live == nil {
		return nil
	}
	records, err := s.live.LoadDeviceAuth()
	if err != nil {
		return err
	}
	for _, rec := range records {
		if rec == nil || rec.DeviceID == "" {
			continue
		}
		s.devices[rec.DeviceID] = rec
		if rec.DeviceToken != "" {
			s.tokens[rec.DeviceToken] = rec.DeviceID
		}
	}
	return nil
}

func splitLines(b []byte) [][]byte {
	var lines [][]byte
	start := 0
	for i, c := range b {
		if c == '\n' {
			lines = append(lines, b[start:i])
			start = i + 1
		}
	}
	if start < len(b) {
		lines = append(lines, b[start:])
	}
	return lines
}

func (s *Store) persistDevices() error {
	if s.disableDisk {
		return nil
	}
	list := make([]*deviceRecord, 0, len(s.devices))
	for _, d := range s.devices {
		list = append(list, d)
	}
	data, err := json.MarshalIndent(list, "", "  ")
	if err != nil {
		return err
	}
	return os.WriteFile(filepath.Join(s.dataDir, "devices.json"), append(data, '\n'), 0o600)
}

func (s *Store) appendEvent(ev models.Event) error {
	if s.disableDisk {
		return nil
	}
	f, err := os.OpenFile(filepath.Join(s.dataDir, "events.jsonl"), os.O_APPEND|os.O_CREATE|os.O_WRONLY, 0o600)
	if err != nil {
		return err
	}
	defer f.Close()
	data, err := json.Marshal(ev)
	if err != nil {
		return err
	}
	_, err = f.Write(append(data, '\n'))
	return err
}

func (s *Store) Register(req models.RegisterRequest) (*models.RegisterResponse, error) {
	s.mu.Lock()
	defer s.mu.Unlock()

	deviceID := req.DeviceID
	if deviceID == "" {
		deviceID = idgen.NewDeviceID()
	}
	rec, ok := s.devices[deviceID]
	if !ok {
		rec = &deviceRecord{DeviceID: deviceID}
		s.devices[deviceID] = rec
	}
	if rec.DeviceToken == "" {
		rec.DeviceToken = idgen.NewToken()
		s.tokens[rec.DeviceToken] = deviceID
	}
	now := s.clock.Now()
	rec.LastSeenAt = &now
	if rec.LastDetails == nil {
		rec.LastDetails = map[string]any{}
	}
	if req.Hostname != "" {
		rec.LastDetails["hostname"] = req.Hostname
	}
	if req.OS != "" {
		rec.LastDetails["os"] = req.OS
	}
	if req.OSVersion != "" {
		rec.LastDetails["os_version"] = req.OSVersion
	}
	if req.Arch != "" {
		rec.LastDetails["arch"] = req.Arch
	}
	if req.AgentVersion != "" {
		rec.LastDetails["agent_version"] = req.AgentVersion
	}
	if err := s.persistDevices(); err != nil {
		return nil, err
	}
	if s.live != nil {
		s.live.UpsertRegister(deviceID, rec.LastDetails, now)
		if s.disableDisk {
			if err := s.live.SaveDeviceAuth(rec); err != nil {
				return nil, err
			}
		}
	}
	return &models.RegisterResponse{DeviceID: deviceID, DeviceToken: rec.DeviceToken}, nil
}

func (s *Store) DeviceIDForToken(token string) (string, bool) {
	s.mu.RLock()
	defer s.mu.RUnlock()
	id, ok := s.tokens[token]
	return id, ok
}

func (s *Store) AddEvent(ev models.Event) error {
	s.mu.Lock()
	defer s.mu.Unlock()

	rec, ok := s.devices[ev.DeviceID]
	if !ok {
		rec = &deviceRecord{DeviceID: ev.DeviceID}
		s.devices[ev.DeviceID] = rec
	}
	now := ev.TS
	if now.IsZero() {
		now = s.clock.Now()
		ev.TS = now
	}
	rec.LastSeenAt = &now
	if ev.Type == constants.TypeClientDetails {
		rec.LastDetails = ev.Payload
	}
	list := append(s.events[ev.DeviceID], ev)
	if len(list) > s.maxEvents {
		list = list[len(list)-s.maxEvents:]
	}
	s.events[ev.DeviceID] = list
	if err := s.appendEvent(ev); err != nil {
		return err
	}
	if err := s.persistDevices(); err != nil {
		return err
	}
	if s.live != nil {
		s.live.UpsertEvent(ev)
		if s.disableDisk {
			if err := s.live.SaveDeviceAuth(rec); err != nil {
				return err
			}
		}
	}
	if s.publisher != nil {
		s.publisher.PublishEvent(ev)
	}
	return nil
}

func (s *Store) GetClient(deviceID string, limit int) (*models.ClientView, bool) {
	s.mu.RLock()
	defer s.mu.RUnlock()
	rec, ok := s.devices[deviceID]
	if !ok {
		return nil, false
	}
	if limit <= 0 {
		limit = 50
	}
	events := s.events[deviceID]
	if len(events) > limit {
		events = events[len(events)-limit:]
	}
	outEvents := make([]models.Event, len(events))
	copy(outEvents, events)
	return &models.ClientView{
		DeviceID:     rec.DeviceID,
		LastDetails:  rec.LastDetails,
		LastSeenAt:   rec.LastSeenAt,
		RecentEvents: outEvents,
	}, true
}
